handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java

```diff
@@ -1318,6 +1295,7 @@ private HandshakeStatus handshake() throws SSLException {
         }
         // if SSL_do_handshake returns > 0 or sslError == SSL.SSL_ERROR_NAME it means the handshake was finished.
         session.handshakeFinished();
+        engineMap.remove(ssl);
```

Comment:
Follow up investigation: I wonder if we can get rid of `engineMap` by storing a reference to the `OpenSslEngine` in the native code (ssl user data?). It would simplify things to not have to map from `long` to `OpenSslEngine` in the java land ... and also remove a concurrent map.